File: bindings/elixir/lib/proven/safe_color.ex
# SPDX-License-Identifier: PMPL-1.0-or-later
# Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

defmodule Proven.SafeColor do
  @moduledoc """
  Safe color manipulation via libproven FFI.

  Color parsing, RGB/HSL conversion, and hex formatting are performed
  by the Idris 2 verified core.
  """

  @type rgb :: {0..255, 0..255, 0..255}
  @type hsl :: {float(), float(), float()}

  @doc """
  Parse a hex color string to RGB tuple.

  Supports formats: #RGB, #RRGGBB, RGB, RRGGBB.

  Returns `{:ok, {r, g, b}}` or `{:error, :invalid_format}`.
  """
  @spec from_hex(String.t()) :: {:ok, rgb()} | {:error, atom()}
  def from_hex(hex) when is_binary(hex) do
    case Proven.NIF.nif_color_parse_hex(hex) do
      {:ok, {r, g, b}} -> {:ok, {r, g, b}}
      {:error, _} -> {:error, :invalid_format}
    end
  end

  @doc """
  Convert RGB to hex string.

  Returns `{:ok, hex_string}` or `{:error, reason}`.
  """
  @spec to_hex(rgb()) :: {:ok, String.t()} | {:error, atom()}
  def to_hex({r, g, b})
      when is_integer(r) and r >= 0 and r <= 255 and
             is_integer(g) and g >= 0 and g <= 255 and
             is_integer(b) and b >= 0 and b <= 255 do
    Proven.NIF.nif_color_to_hex(r, g, b)
  end

  def to_hex(_), do: {:error, :invalid_rgb}

  @doc """
  Convert RGB to HSL.

  Returns `{:ok, {h, s, l}}` where h is 0-360, s and l are 0.0-1.0.
  """
  @spec to_hsl(rgb()) :: {:ok, hsl()}
  def to_hsl({r, g, b})
      when is_integer(r) and r >= 0 and r <= 255 and
             is_integer(g) and g >= 0 and g <= 255 and
             is_integer(b) and b >= 0 and b <= 255 do
    {h, s, l} = Proven.NIF.nif_color_rgb_to_hsl(r, g, b)
    {:ok, {h, s, l}}
  end

  def to_hsl(_), do: {:error, :invalid_rgb}
end

File: bindings/haskell/src/Proven/SafeColor.hs
{- SPDX-License-Identifier: PMPL-1.0-or-later -}
{- Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk> -}

-- | Safe color operations via libproven FFI.
--
-- Color parsing and conversion are performed by the Idris 2 verified core.
module Proven.SafeColor
  ( RGB(..)
  , HSL(..)
  , parseHexColor
  , rgbToHsl
  , colorToHex
  ) where

import Data.Word (Word8)
import Proven.FFI (c_proven_color_parse_hex, c_proven_color_rgb_to_hsl,
                   c_proven_color_to_hex, c_proven_free_string)
import Proven.FFI.Types (RGBColor(..), HSLColor(..), ColorParseResult(..))
import Proven.Core (withCStringLen', stringResultToMaybe)

-- | RGB color with 8-bit components.
data RGB = RGB
  { rgbRed   :: !Word8
  , rgbGreen :: !Word8
  , rgbBlue  :: !Word8
  } deriving (Eq, Show)

-- | HSL color (hue 0-360, saturation 0-1, lightness 0-1).
data HSL = HSL
  { hslHue        :: !Double
  , hslSaturation :: !Double
  , hslLightness  :: !Double
  } deriving (Eq, Show)

-- | Parse a hex color string (#RRGGBB or #RGB).
-- Delegates to @proven_color_parse_hex@ in libproven.
parseHexColor :: String -> IO (Maybe RGB)
parseHexColor str = withCStringLen' str $ \ptr len -> do
  result <- c_proven_color_parse_hex ptr len
  if cprStatusRaw result == 0
    then let c = cprColor result
         in return (Just (RGB (rgbR c) (rgbG c) (rgbB c)))
    else return Nothing

-- | Convert RGB to HSL.
-- Delegates to @proven_color_rgb_to_hsl@ in libproven.
rgbToHsl :: RGB -> IO HSL
rgbToHsl (RGB r g b) = do
  result <- c_proven_color_rgb_to_hsl (RGBColor r g b)
  return (HSL (realToFrac (hslH result))
              (realToFrac (hslS result))
              (realToFrac (hslL result)))

-- | Format RGB color as hex string.
-- Delegates to @proven_color_to_hex@ in libproven.
colorToHex :: RGB -> IO (Maybe String)
colorToHex (RGB r g b) = do
  sr <- c_proven_color_to_hex (RGBColor r g b)
  stringResultToMaybe c_proven_free_string sr
